federation/pkg/federatedtypes/hpa.go

```diff
@@ -0,0 +1,876 @@
+	}
+	hpa2.Spec.MaxReplicas = hpa1.Spec.MaxReplicas
+	return fedutil.ObjectMetaAndSpecEquivalent(hpa1, hpa2)
+}
+
+type replicaNums struct {
+	min int32
+	max int32
+}
+
+type hpaFederatedStatus struct {
+	lastScaleTime *metav1.Time
+	// Indicates how many clusters have hpa/replicas.
+	// Used to average the cpu utilization which is
+	// reflected to the federation user.
+	count                           int32
+	currentCPUUtilizationPercentage *int32
+	currentReplicas                 int32
+	desiredReplicas                 int32
+}
+
+type hpaSchedulingInfo struct {
+	scheduleState map[string]*replicaNums
+	fedStatus     hpaFederatedStatus
+}
+
+// List of cluster names.
+type hpaLists struct {
+	// Stores names of those clusters which can offer min.
+	availableMin sets.String
+	// Stores names of those clusters which can offer max.
+	availableMax sets.String
+	// Stores names of those clusters which do not have hpa yet.
+	noHpa sets.String
+}
+
+func (a *HpaAdapter) GetSchedule(obj pkgruntime.Object, key string, clusters []*federationapi.Cluster, informer fedutil.FederatedInformer) (interface{}, error) {
+	currentClusterObjs, err := getCurrentClusterObjs(informer, key, clusters)
+	if err != nil {
+		return nil, err
+	}
+
+	// Initialise averaged cpu utilisation for this reconcile.
+	var ccup int32 = 0
+	fedStatus := hpaFederatedStatus{
+		currentCPUUtilizationPercentage: &ccup,
+		count:           int32(0),
+		desiredReplicas: int32(0),
+		currentReplicas: int32(0),
+	}
+	fedHpa := obj.(*autoscalingv1.HorizontalPodAutoscaler)
+	// We assign the last known scale time here, which we update with
+	// the latest time from among all clusters in ScheduleObject()
+	if fedHpa.Status.LastScaleTime != nil {
+		t := metav1.NewTime(fedHpa.Status.LastScaleTime.Time)
+		fedStatus.lastScaleTime = &t
+	}
+
+	return &hpaSchedulingInfo{
+		scheduleState: getHpaScheduleState(obj, currentClusterObjs),
+		fedStatus:     fedStatus,
+	}, nil
+}
+
+func getCurrentClusterObjs(informer fedutil.FederatedInformer, key string, clusters []*federationapi.Cluster) (map[string]pkgruntime.Object, error) {
+	currentClusterObjs := make(map[string]pkgruntime.Object)
+	for _, cluster := range clusters {
+		clusterName := cluster.Name
+		clusterObj, found, err := informer.GetTargetStore().GetByKey(clusterName, key)
+		if err != nil {
+			return nil, err
+		}
+		currentClusterObjs[clusterName] = nil
+		if found {
+			currentClusterObjs[clusterName] = clusterObj.(pkgruntime.Object)
+		}
+	}
+	return currentClusterObjs, nil
+}
+
+func getHpaScheduleState(fedObj pkgruntime.Object, currentObjs map[string]pkgruntime.Object) map[string]*replicaNums {
+	fedHpa := fedObj.(*autoscalingv1.HorizontalPodAutoscaler)
+	requestedMin := int32(1)
+	if fedHpa.Spec.MinReplicas != nil {
+		requestedMin = *fedHpa.Spec.MinReplicas
+	}
+	requestedReplicas := replicaNums{
+		min: requestedMin,
+		max: fedHpa.Spec.MaxReplicas,
+	}
+	// replica distribution count, per cluster
+	rdc := replicaNums{
```

Comment:
When I was at including cluster weights for the hpa (using annotations on fed object). I could not really contemplate, how/what a user might want the weights to be put use to. 
I mean autoscaler is already specifying the target cpu utilization, and thats the criteria we choose to do an eventual distribution of hpa replicas across clusters, to an effect that the replicas remain or move where they are needed most. I could think of specific min/max which could be hard limits put by user per cluster (which I haven't implemented yet and intend to do as a follow up). I could as mentioned not figure out real usage of weights put on hpas (as to what to apply weights to, cpu utilisation?). I see here that you are suggesting that the replica distribution number could be weighted (which I see as contradicting use as it can eventually conflict with the conditions we use to move replicas around, which for the autoscaler target cpu utilization). I think more intricate discussion is needed on this and would request that we let it be as a follow up (I see hpa clearly usable without preferences also). Have put an item in the follow up issue https://github.com/kubernetes/kubernetes/issues/49644.